pkg/cloudprovider/gce/gce.go

```diff
@@ -421,6 +464,57 @@ func (gce *GCECloud) GetNodeResources(name string) (*api.NodeResources, error) {
 	}
 }
 
+func getMetadataValue(metadata *compute.Metadata, key string) string {
+	for _, item := range metadata.Items {
+		if item.Key == key {
+			return item.Value
+		}
+	}
+	return ""
+}
+
+func (gce *GCECloud) Configure(name string, spec *api.NodeSpec) error {
+	instanceName := canonicalizeInstanceName(name)
+	instanceCall := gce.service.Instances.Get(gce.projectID, gce.zone, instanceName)
+	instance, err := instanceCall.Do()
+	if err != nil {
+		return err
+	}
+	if currentValue := getMetadataValue(instance.Metadata, podCIDRMetadataKey); currentValue != "" {
+		if currentValue == spec.PodCIDR {
+			// IP range already set to proper value.
+			return nil
+		}
+		return ErrMetadataConflict
+	}
+	instance.Metadata.Items = append(instance.Metadata.Items,
+		&compute.MetadataItems{
+			Key:   podCIDRMetadataKey,
+			Value: spec.PodCIDR,
+		})
+	setMetadataCall := gce.service.Instances.SetMetadata(gce.projectID, gce.zone, instanceName, instance.Metadata)
+	setMetadataOp, err := setMetadataCall.Do()
+	if err != nil {
+		return err
+	}
+	err = gce.waitForZoneOp(setMetadataOp)
+	if err != nil {
+		return err
+	}
+	insertCall := gce.service.Routes.Insert(gce.projectID, &compute.Route{
+		Name:            instanceName,
+		DestRange:       spec.PodCIDR,
+		NextHopInstance: fmt.Sprintf("zones/%s/instances/%s", gce.zone, instanceName),
+		Network:         fmt.Sprintf("global/networks/%s", gce.networkName),
+		Priority:        1000,
+	})
+	insertOp, err := insertCall.Do()
+	if err != nil {
+		return err
+	}
+	return gce.waitForGlobalOp(insertOp)
```

Comment:
Serializing metadata reads/writes and route inserts for each node is going to be horribly slow and wasteful of quota.

I'd suggest doing the metadata read/writes as a batch: Read metadata once, append all new CIDR keys, write back once.
Then kick off a goroutine to insert a route and wait on the op for each node.

This will require a little bit of restructuring up in SyncCloud.